src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+  const time_t abort_time =
+      absl::ToTimeT(absl::FromUnixSeconds(presumed_abort_time.seconds()) +
+                    absl::Nanoseconds(presumed_abort_time.nanos()));
+  return Blockchain::StartVoting(transaction_id, abort_time, num_cohorts);
+}
+
+absl::StatusOr<Blockchain::VotingDecision> CoordinatorServer::GetVotingDecision(
+    const std::string &transaction_id) {
+  return Blockchain::GetVotingDecision(transaction_id);
+}
+
+absl::Time CoordinatorServer::Now() { return absl::Now(); }
+
+void CoordinatorServer::PrepareCohortTransaction(
+    const std::string &transaction_id, const Namespace &ns,
+    const cohort::PrepareTransactionRequest &request,
+    const grpc::ServerContext &context) {
+  metadata_by_transaction_[transaction_id].contexts.push_back(
+      ClientContext::FromServerContext(context));
+  GetCohortStub(ns).AsyncPrepareTransaction(
+      metadata_by_transaction_[transaction_id].contexts.back().get(), request,
+      /*cq = */ nullptr);
+}
+
+std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+    cohort::GetTransactionResultResponse>>
+CoordinatorServer::AsyncGetResultsFromCohort(
+    const Namespace &ns, const cohort::GetTransactionResultRequest &request,
+    ClientContext &context) {
+  return GetCohortStub(ns).AsyncGetTransactionResult(&context, request,
+                                                     /*cq=*/nullptr);
+}
+
+grpc::Status CoordinatorServer::FinishAsyncGetResultsFromCohort(
+    std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+        cohort::GetTransactionResultResponse>> &async_response,
+    cohort::GetTransactionResultResponse &response) {
+  grpc::Status status;
+  async_response->Finish(&response, &status, /*tag=*/nullptr);
+  return status;
+}
+
+CohortStub &CoordinatorServer::GetCohortStub(const Namespace &ns) {
+  std::unique_ptr<CohortStub> &cohort_stub = cohort_by_namespace_[ns.address()];
+  if (!cohort_stub) {
+    cohort_stub = cohort::Cohort::NewStub(
+        grpc::CreateChannel(ns.address(), grpc::InsecureChannelCredentials()));
+  }
+  return *cohort_stub;
+}
+
+grpc::Status CoordinatorServer::UpdateResponseForSingleCohortTransaction(
+    const std::string &transaction_id, const ServerContext &context) {
+  GetTransactionResultResponse &response =
+      response_by_transaction_[transaction_id].response;
+  cohort::GetTransactionResultRequest cohort_request;
+  cohort_request.set_transaction_id(transaction_id);
+  std::vector<std::unique_ptr<ClientContext>> transaction_result_contexts;
+  transaction_result_contexts.push_back(
+      ClientContext::FromServerContext(context));
+  cohort::GetTransactionResultResponse cohort_response;
+  const Namespace &ns =
+      metadata_by_transaction_[transaction_id].single_cohort_namespace.value();
+  auto async_response = AsyncGetResultsFromCohort(
+      ns, cohort_request, *transaction_result_contexts.back());
+  grpc::Status status =
+      FinishAsyncGetResultsFromCohort(async_response, cohort_response);
```

Comment:
Why do you want to use async call here when there's only a single cohort? Are you avoiding to block the main thread here?

Reply:
Changed to a synchronous call.